### ISSUE TITLE
Intermittent NULL pointer dereference in p_arch_static_call_transform_entry() on unload of other modules

### ISSUE THREAD

**Issue by solardiz** [openwall/lkrg#85]
```
[  OK  ] Stopped Remount Root and Kernel File Systems.
[  OK  ] Reached target Shutdown.
[  OK  ] Reached target Final Step.
[  OK  ] Finished Power-Off.
[  OK  ] Reached target Power-Off.
[   41.384711] [p_lkrg] [TRACEPOINT] Updated module's list hash for module[virtio_dma_buf] but can't find the same module in KOBJs list!
[   41.438382] [p_lkrg] [TRACEPOINT] Updated module's list hash for module[virtio_dma_buf] but can't find the same module in KOBJs list!
[   41.462114] BUG: kernel NULL pointer dereference, address: 0000000000000000
[   41.465171] #PF: supervisor read access in kernel mode
[   41.465171] #PF: error_code(0x0000) - not-present page
[   41.465171] PGD 0 P4D 0 
[   41.465171] Oops: 0000 [#1] SMP NOPTI
[   41.465171] CPU: 0 PID: 404 Comm: systemd-udevd Tainted: G           OE     5.11.0-16-generic #17-Ubuntu
[   41.465171] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.13.0-1ubuntu1.1 04/01/2014
[   41.465171] RIP: 0010:p_arch_static_call_transform_entry+0xed/0x180 [p_lkrg]
[   41.465171] Code: c1 48 85 c0 0f 84 83 00 00 00 8b 35 f5 34 02 00 85 f6 74 cc 48 8b 15 f2 34 02 00 31 c0 eb 0b 83 c0 01 48 83 c2 60 39 c6 74 b6 <48> 3b 0a 75 f0 89 05 80 34 02 00 eb a9 48 8b 05 67 f8 01 00 4c 89
[   41.465171] RSP: 0018:ffffb639803f7b98 EFLAGS: 00000246
[   41.465171] RAX: 0000000000000000 RBX: ffff9f373530e280 RCX: ffffffffc074c8c0
[   41.465171] RDX: 0000000000000000 RSI: 000000000000001e RDI: ffffffffc06cab4a
[   41.465171] RBP: ffffb639803f7ba8 R08: 0000000000000000 R09: 00000000000000b6
[   41.465171] R10: ffff9f3702cef680 R11: 0000000000000012 R12: ffffffffc06cab4a
[   41.465171] R13: 0000000000000000 R14: ffffffffc03f1a38 R15: ffff9f373ec18440
[   41.465171] FS:  00007feb241f68c0(0000) GS:ffff9f373ec00000(0000) knlGS:0000000000000000
[   41.465171] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   41.465171] CR2: 0000000000000000 CR3: 00000000058a8000 CR4: 00000000000006f0
[   41.465171] Call Trace:
[   41.465171]  pre_handler_kretprobe+0x97/0x170
[   41.465171]  ? arch_static_call_transform+0x1/0x90
[   41.465171]  kprobe_ftrace_handler+0x11e/0x1f0
[   41.465171]  ? arch_static_call_transform+0x5/0x90
[   41.465171]  ? kvm_dirty_ring_push+0x6a/0x80 [kvm]
[   41.465171]  0xffffffffc03bb0e3
[   41.465171] RIP: 0010:arch_static_call_transform+0x1/0x90
[   41.465171] Code: 55 48 89 fa 48 c7 c7 20 fd 99 83 c6 05 94 0b ed 01 01 48 89 e5 e8 5c b3 b9 00 0f 0b 5d c3 c3 66 2e 0f 1f 84 00 00 00 00 00 e8 <5b> b2 f7 3d 55 48 89 e5 41 56 49 89 fe 48 c7 c7 20 b4 06 84 41 55
[   41.465171] RSP: 0018:ffffb639803f7d08 EFLAGS: 00000292 ORIG_RAX: 0000000000000000
[   41.465171] RAX: ffff9f3702cef680 RBX: ffffffffc074c8c0 RCX: 0000000000000000
[   41.465171] RDX: ffffffffc06bc540 RSI: 0000000000000000 RDI: ffffffffc06cab4a
[   41.465171] RBP: ffffb639803f7d48 R08: 0000000000000020 R09: ffffffffc0737890
[   41.465171] R10: ffff9f3702cef680 R11: 0000000000000012 R12: ffffffffc06cab4a
[   41.465171] R13: ffffffffc0737890 R14: ffffffffc074c2c4 R15: ffffffffc074c2c0
[   41.465171]  ? kvm_dirty_ring_push+0x6a/0x80 [kvm]
[   41.465171]  ? __traceiter_kvm_halt_poll_ns+0x60/0x60 [kvm]
[   41.465171]  ? kvm_dirty_ring_push+0x6a/0x80 [kvm]
[   41.465171]  ? arch_static_call_transform+0x5/0x90
[   41.465171]  ? __static_call_init.part.0+0x16c/0x220
[   41.465171]  ? arch_static_call_transform+0x5/0x90
[   41.465171]  ? __static_call_init.part.0+0x16c/0x220
[   41.465171]  static_call_add_module+0xea/0x110
[   41.465171]  static_call_module_notify+0x52/0xc0
[   41.465171]  blocking_notifier_call_chain_robust+0x6a/0xe0
[   41.465171]  ? klp_module_coming+0xe6/0x110
[   41.465171]  load_module+0x400/0x780
[   41.465171]  __do_sys_finit_module+0xc2/0x120
[   41.465171]  __x64_sys_finit_module+0x1a/0x20
[   41.465171]  do_syscall_64+0x38/0x90
[   41.465171]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[   41.465171] RIP: 0033:0x7feb246aef6d
[   41.465171] Code: 28 0d 00 0f 05 eb a9 66 0f 1f 44 00 00 f3 0f 1e fa 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d cb de 0c 00 f7 d8 64 89 01 48
[   41.465171] RSP: 002b:00007ffcd4e6b108 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[   41.465171] RAX: ffffffffffffffda RBX: 000056381a8947e0 RCX: 00007feb246aef6d
[   41.465171] RDX: 0000000000000000 RSI: 00007feb24854e2d RDI: 0000000000000006
[   41.465171] RBP: 0000000000020000 R08: 0000000000000000 R09: 000056381a8947e0
[   41.465171] R10: 0000000000000006 R11: 0000000000000246 R12: 00007feb24854e2d
[   41.465171] R13: 0000000000000000 R14: 000056381a897660 R15: 000056381a8947e0
[   41.465171] Modules linked in: kvm(+) virtio_gpu virtio_dma_buf drm_kms_helper input_leds i2c_i801 psmouse cec rc_core fb_sys_fops syscopyarea i2c_smbus sysfillrect sysimgblt mac_hid lpc_ich sch_fq_codel drm ip_tables x_tables ahci virtio_scsi libahci serio_raw qemu_fw_cfg p_lkrg(OE) dm_mirror dm_region_hash dm_log virtio_rng autofs4
[   41.465171] CR2: 0000000000000000
[   41.510307] ---[ end trace 585af7df4cccec66 ]---
[   41.510653] RIP: 0010:p_arch_static_call_transform_entry+0xed/0x180 [p_lkrg]
[   41.511146] Code: c1 48 85 c0 0f 84 83 00 00 00 8b 35 f5 34 02 00 85 f6 74 cc 48 8b 15 f2 34 02 00 31 c0 eb 0b 83 c0 01 48 83 c2 60 39 c6 74 b6 <48> 3b 0a 75 f0 89 05 80 34 02 00 eb a9 48 8b 05 67 f8 01 00 4c 89
[   41.512307] RSP: 0018:ffffb639803f7b98 EFLAGS: 00000246
[   41.512670] RAX: 0000000000000000 RBX: ffff9f373530e280 RCX: ffffffffc074c8c0
[   41.513071] RDX: 0000000000000000 RSI: 000000000000001e RDI: ffffffffc06cab4a
[   41.513767] RBP: ffffb639803f7ba8 R08: 0000000000000000 R09: 00000000000000b6
[   41.514247] R10: ffff9f3702cef680 R11: 0000000000000012 R12: ffffffffc06cab4a
[   41.515099] R13: 0000000000000000 R14: ffffffffc03f1a38 R15: ffff9f373ec18440
[   41.515896] FS:  00007feb241f68c0(0000) GS:ffff9f373ec00000(0000) knlGS:0000000000000000
[   41.516302] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   41.516595] CR2: 0000000000000000 CR3: 00000000058a8000 CR4: 00000000000006f0
[   41.517076] Kernel panic - not syncing: Fatal exception
[   41.517153] Kernel Offset: 0x1400000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[   41.517153] ACPI MEMORY or I/O RESET_REG.
```

**Comment by Adam-pi3** [comment]
It looks like I've missed one specific use case and we need an extra synchronization mechanism in `p_arch_static_call_transform_entry`. Per original comment:

```
    * We don't need to take any lock here because there can be only one instance of
    * 'arch_static_call_transform' function running at a time. It's being guarded in
    * a few ways by upper layer via 'static_call_lock', 'tracepoints_mutex' and 'event_mutex'
```

Which is mostly true. However, one function:

`static_call_module_notify -> static_call_add_module -> __static_call_init`

even is running under `static_call_lock` (so it's correctly synchronized) it calls the hooked function under the for() loop:


	for (site = start; site < stop; site++) {
		void *site_addr = static_call_addr(site);
...
...
		arch_static_call_transform(site_addr, NULL, key->func,
				static_call_is_tail(site));
	}

So there isn't any chance for various modules to be racing but 1 module might call static transformation in the loop and race with itself.
It looks like we hit this situation in that bug.

However, our lock must be carefully designed. Most likely similar to our SELinux sync lock logic. The main reason why is because after our entry hook kernel takes `text_mutex` which we rely on in the various places and it also means we can't take just simple spinlock (mutex is out of scope).

There is also one more caveat. This specific situation can only appear when module is not formed yet (COMING state) and we don't care about such modules until they change the state to LIVE. So one of the solution is not to introduce a lock but to detect that we have been called from the function forming the module and just ignore that calls (immediately return and don't do anything). It might be better approach and it won't hurt perf.

Btw. I'm going to be 1 week off starting from tomorrow and won't be able to work on it until the week of 24th of May...

**Comment by solardiz** [comment]
> 1 module might call static transformation in the loop and race with itself.

How would it "race with itself"? Aren't the calls from the loop sequential?

> This specific situation can only appear when module is not formed yet (COMING state) and we don't care about such modules until they change the state to LIVE. So one of the solution is not to introduce a lock but to detect that we have been called from the function forming the module and just ignore that calls

What would happen when the module is eventually LIVE? Would we be called again and handle it appropriately, or would we miss it (and with what consequences)?

**Comment by Adam-pi3** [comment]
> How would it "race with itself"? Aren't the calls from the loop sequential?

They are and I have no idea how that happened. However, if you look at the stack trace it says it's possible:

```
[   41.465171]  ? kvm_dirty_ring_push+0x6a/0x80 [kvm]
**[   41.465171]  ? arch_static_call_transform+0x5/0x90**
[   41.465171]  ? __static_call_init.part.0+0x16c/0x220
**[   41.465171]  ? arch_static_call_transform+0x5/0x90**
[   41.465171]  ? __static_call_init.part.0+0x16c/0x220
[   41.465171]  static_call_add_module+0xea/0x110
```

Also, in our logic of the hook we verify if we won't operate on NULL pointers and later we have NULL deref. This is impossible per se, unless we have another instance running the same logic and zeros temp values.

> What would happen when the module is eventually LIVE? Would we be called again and handle it appropriately, or would we miss it (and with what consequences)?

If module become LIVE, we have separate logic to dump that module in our DB at that specific time (COMING -> LIVE). Later, any static calls will be handle as should be (with the current logic)

**Comment by solardiz** [comment]
The stack trace looks like it's either recursion or a stack trace glitch. To have a race, it'd have been 2+ different execution contexts, and within that one stack trace we see just one. If it's a race, then it's against something else not seen here, not against "itself". My best guess is there's some other place in the kernel (not this `for` loop you found) calling us concurrently. This would imply the comment is actually wrong. I think we need to (1) figure this out and (2) reproduce the problem (perhaps by running a focused stress-test) so that we'd be able to test our eventual fix. Of course, only after your return a week later.

Have a good weekend!

**Comment by Adam-pi3** [comment]
Right. However, I've verified all the calls to `arch_static_call_transform` and there isn't any single one which is not guarded at least by `static_call_lock`. Only in that instance we have loop (multiple calls) to  `arch_static_call_transform` instead of one call. Also take into account that the return code of the call is ignored. My theory is that this function is heavily optimized (and it is) which can be seen from the symbol names (__static_call_init.part.0). Function is split into a few 'parts' and it might be the case that loop can continue to run while `arch_static_call_transform` is being independently executed. E.g. could be that call to `arch_static_call_transform` is handled by another CPU while the caller continues loop iteration.

I would suggest to go to the path of not adding synchronization object but filtering out the calls from not fully formed modules. If that won't help (but I got a feeling it will) we can add then synchronizer as well since regardless of it, filtering calls from not fully formed modules is desired. Although, I'm not sure if that possible. We might do a simple trick by having local to the entry function temporary spinlock, and if any of the src/dst address is pointing into the module which state is not LIVE, unlock the temporary lock and return. Otherwise unlock and continue run. I can add this changes today and we can try to run `mkosi` over the week to see if it helps

**Comment by solardiz** [comment]
> My theory is that this function is heavily optimized

Any such optimization would be within the current execution context.

> Function is split into a few 'parts' and it might be the case that loop can continue to run while `arch_static_call_transform` is being independently executed.

This doesn't make sense to me, unless you're suggesting that a part of the function/loop got moved by the compiler to after release of the lock. That would be a bug in the compiler or in the way the lock primitives are defined (they should be such that the compiler would know not to move stuff across them).

> E.g. could be that call to `arch_static_call_transform` is handled by another CPU while the caller continues loop iteration.

I can't figure out if you suggest this very loop could somehow be parallelized across 2+ CPUs by the compiler (I think that's nonsense) or if you mean the scenario I described above (where the "independently executed" call would also be unrelated to that loop - just something that was permitted to happen elsewhere after release of the lock).

> I would suggest to go to the path of not adding synchronization object but filtering out the calls from not fully formed modules.

I don't mind, but I don't fully understand your proposed implementation (perhaps because I'm not familiar with a few other related things here) and I think that ideally we'd manage to reproduce the problem (but it's fine if you attempt reproducing it on the old code after the week's test).

I suggest that you make these changes via a PR, not directly on main branch, so that we see if `mkosi` is happy with them at least initially (and won't merge if it is not). Thanks!

**Comment by solardiz** [comment]
Just to be clear, my opinion remains this: "My best guess is there's some other place in the kernel (not this for loop you found) calling us concurrently. This would imply the comment is actually wrong."

**Comment by Adam-pi3** [comment]
Pull Request includes locking as well: https://github.com/openwall/lkrg/pull/89

**Comment by Adam-pi3** [comment]
@solardiz it looks like the patch helps. If that's the case I will clean up the PR and we can merge it (if you agree)

**Comment by solardiz** [comment]
> I will clean up the PR and we can merge it (if you agree)

Yes, please do (and while you're at it, perhaps also update the comment on locking). Thanks!

**Comment by solardiz** [comment]
Looks like we got more of these crashes in https://github.com/openwall/lkrg/runs/2693541239

**Comment by Adam-pi3** [comment]
Let see if we still encounter that problem after https://github.com/openwall/lkrg/commit/9c592c662041dfd642cb915920917c07773cb81a

**Comment by Adam-pi3** [comment]
Did we see any case of the described problem or we can close this issue?
CC: @solardiz @vt-alt

**Comment by solardiz** [comment]
> Did we see any case of the described problem or we can close this issue?

No such cases recently. Let's close. We can reopen if necessary.
